File: kmeans_HW1.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MAX_ITER 1000
#define DEFAULT_ITER 200
#define EPSILON 0.001
#define MAX_LINE_LENGTH 1000
#define MAX_DOUBLE 1.7976931348623157E+308

typedef struct {
    double* coordinates;
    int dimension;
    struct Cluster* cluster;
} Point;

typedef struct Cluster {
    Point centroid;
    Point sum_of_points;
    int number_of_points;
    int is_smaller_than_epsilon;
} Cluster;

/* Function prototypes */
Point* read_points_from_stdin(int* num_points, int* dimension);
void init_clusters(Cluster* clusters, Point* datapoints, int k, int dimension);
double distance_between_points(Point* p1, Point* p2);
void assign_point_to_cluster(Point* point, Cluster* cluster);
double update_centroid(Cluster* cluster);
Cluster* find_nearest_cluster(Point* point, Cluster* clusters, int k);
void free_memory(Point* datapoints, int num_points, Cluster* clusters, int k);
double parse_double(char** str);

int main(int argc, char* argv[]) {
    int k, max_iter, num_points, dimension;
    int iteration, done_clusters, i, j;
    Point* datapoints;
    Cluster* clusters;
    Cluster* nearest_cluster;

    if (argc < 2 || argc > 3) {
        printf("An Error Has Occured");
        return 1;
    }

    k = atoi(argv[1]);
    max_iter = (argc == 3) ? atoi(argv[2]) : DEFAULT_ITER;
    
    if (k <= 1 || k >= 1000) {
        printf("Invalid number of clusters!\n");
        return 1;
    }
    
    if (max_iter <= 1 || max_iter >= MAX_ITER) {
        printf("Invalid maximum iteration!\n");
        return 1;
    }

    num_points = 0;
    dimension = 0;
    datapoints = read_points_from_stdin(&num_points, &dimension);
    
    if (k >= num_points) {
        printf("Invalid number of clusters!\n");
        free_memory(datapoints, num_points, NULL, 0);
        return 1;
    }

    clusters = (Cluster*) malloc(k * sizeof(Cluster));
    if (!clusters) {
        printf("An Error Has Occured\n");
        free_memory(datapoints, num_points, NULL, 0);
        return 1;
    }

    init_clusters(clusters, datapoints, k, dimension);

    iteration = 0;
    done_clusters = 0;

    while (iteration < max_iter && done_clusters < k) {
        done_clusters = 0;

        for (i = 0; i < num_points; i++) {
            nearest_cluster = find_nearest_cluster(&datapoints[i], clusters, k);
            assign_point_to_cluster(&datapoints[i], nearest_cluster);
        }

        for (i = 0; i < k; i++) {
            double centroid_diff = update_centroid(&clusters[i]);
            if (centroid_diff < EPSILON) {
                done_clusters++;
            }
        }

        iteration++;
    }

    /* Print final centroids */
    for (i = 0; i < k; i++) {
        for (j = 0; j < dimension; j++) {
            printf("%.4f", clusters[i].centroid.coordinates[j]);
            if (j < dimension - 1) printf(",");
        }
        printf("\n");
    }

    free_memory(datapoints, num_points, clusters, k);
    return 0;
}

Point* read_points_from_stdin(int* num_points, int* dimension) {
    char line[MAX_LINE_LENGTH];
    Point* points = NULL;
    int dim;
    double* coords;
    char* line_ptr;

    *num_points = 0;
    *dimension = 0;

    while (fgets(line, MAX_LINE_LENGTH, stdin) != NULL) {
        if (line[0] != '\n') {  /* Ignore empty lines */
            (*num_points)++;
            points = (Point*) realloc(points, *num_points * sizeof(Point));
            if (!points) {
                printf("An Error Has Occurred\n");
                exit(1);
            }

            line_ptr = line;
            dim = 0;
            coords = NULL;

            while (*line_ptr != '\n' && *line_ptr != '\0') {
                dim++;
                coords = (double*) realloc(coords, dim * sizeof(double));
                if (!coords) {
                    printf("An Error Has Occurred\n");
                    exit(1);
                }
                coords[dim - 1] = parse_double(&line_ptr);
                if (*line_ptr == ',') line_ptr++; /* Skip comma */
            }

            if (*dimension == 0) {
                *dimension = dim;
            } else if (dim != *dimension) {
                printf("An Error Has Occurred\n");
                exit(1);
            }

            points[*num_points - 1].coordinates = coords;
            points[*num_points - 1].dimension = dim;
            points[*num_points - 1].cluster = NULL;
        }
    }

    return points;
}

double parse_double(char** str) {
    double result = 0.0;
    int sign = 1;
    int decimal_places = 0;
    int is_decimal = 0;

    /* Handle sign */
    if (**str == '-') {
        sign = -1;
        (*str)++;
    } 
    
    /* Parse integer part */
    while (**str >= '0' && **str <= '9') {
        result = result * 10.0 + (**str - '0');
        (*str)++;
    }

    /* Parse decimal part */
    if (**str == '.') {
        (*str)++;
        is_decimal = 1;
        while (**str >= '0' && **str <= '9') {
            result = result * 10.0 + (**str - '0');
            decimal_places++;
            (*str)++;
        }
    }

    /* Apply decimal point */
    if (is_decimal) {
        while (decimal_places > 0) {
            result /= 10.0;
            decimal_places--;
        }
    }

    return sign * result;
}

void init_clusters(Cluster* clusters, Point* datapoints, int k, int dimension) {
    int i, j;
    for (i = 0; i < k; i++) {
        clusters[i].centroid.coordinates = (double*) malloc(dimension * sizeof(double));
        clusters[i].sum_of_points.coordinates = (double*) malloc(dimension * sizeof(double));
        if (!clusters[i].centroid.coordinates || !clusters[i].sum_of_points.coordinates) {
            printf("An Error Has Occured\n");
            exit(1);
        }

        for (j = 0; j < dimension; j++) {
            clusters[i].centroid.coordinates[j] = datapoints[i].coordinates[j];
            clusters[i].sum_of_points.coordinates[j] = 0.0;
        }

        clusters[i].centroid.dimension = dimension;
        clusters[i].sum_of_points.dimension = dimension;
        clusters[i].number_of_points = 0;
        clusters[i].is_smaller_than_epsilon = 0;
    }
}

double distance_between_points(Point* p1, Point* p2) {
    double sum = 0.0;
    int i;
    for (i = 0; i < p1->dimension; i++) {
        double diff = p1->coordinates[i] - p2->coordinates[i];
        sum += diff * diff;
    }
    return sqrt(sum);
}

void assign_point_to_cluster(Point* point, Cluster* cluster) {
    int i;
    if (point->cluster) {
        point->cluster->number_of_points--;
        for (i = 0; i < point->dimension; i++) {
            point->cluster->sum_of_points.coordinates[i] -= point->coordinates[i];
        }
    }

    point->cluster = cluster;
    cluster->number_of_points++;
    for (i = 0; i < point->dimension; i++) {
        cluster->sum_of_points.coordinates[i] += point->coordinates[i];
    }
}

double update_centroid(Cluster* cluster) {
    double max_diff = 0.0;
    int i;
    for (i = 0; i < cluster->centroid.dimension; i++) {
        double old_coord = cluster->centroid.coordinates[i];
        double diff;
        cluster->centroid.coordinates[i] = cluster->sum_of_points.coordinates[i] / cluster->number_of_points;
        diff = fabs(cluster->centroid.coordinates[i] - old_coord);
        if (diff > max_diff) {
            max_diff = diff;
        }
    }
    cluster->is_smaller_than_epsilon = (max_diff < EPSILON);
    return max_diff;
}

Cluster* find_nearest_cluster(Point* point, Cluster* clusters, int k) {
    double min_distance = MAX_DOUBLE;
    Cluster* nearest_cluster = NULL;
    int i;

    for (i = 0; i < k; i++) {
        double distance = distance_between_points(point, &clusters[i].centroid);
        if (distance < min_distance) {
            min_distance = distance;
            nearest_cluster = &clusters[i];
        }
    }

    return nearest_cluster;
}

void free_memory(Point* datapoints, int num_points, Cluster* clusters, int k) {
    int i;
    for (i = 0; i < num_points; i++) {
        free(datapoints[i].coordinates);
    }
    free(datapoints);

    if (clusters) {
        for (i = 0; i < k; i++) {
            free(clusters[i].centroid.coordinates);
            free(clusters[i].sum_of_points.coordinates);
        }
        free(clusters);
    }
}

File: kmeans_HW1.py
import sys
from math import dist

MAX_ITER = 1000
DEFAULT_ITER = 200
EPSILON = 0.001


class Point:
	coordinates: list
	cluster: None

	def __init__(self, coordinates: list[float]):
		self.coordinates = coordinates.copy()
		self.cluster = None

	def __add__(self, other):
		return [sum(x) for x in zip(self.coordinates, other.coordinates)]

	def __iadd__(self, other):
		self.coordinates = [sum(x) for x in zip(self.coordinates, other.coordinates)]
		return self

	def __isub__(self, other):
		self.coordinates = [x[0] - x[1] for x in zip(self.coordinates, other.coordinates)]
		return self

	def __repr__(self):
		return ",".join(f"{x:0.4f}" for x in self.coordinates)

	def __str__(self):
		return ",".join(f"{x:0.4f}" for x in self.coordinates)

	def __copy__(self):
		return Point(coordinates=self.coordinates)

	# Assigns/Unassigns a point to/from a cluster
	def assign_to_cluster(self, cluster) -> None:
		if self.cluster:
			self.cluster.remove_point(point=self)

		cluster.add_point(point=self)
		self.cluster = cluster

	# Distance between 2 points
	def distance_to_point(self, other_point) -> float:
		return dist(self.get_coordinates(), other_point.get_coordinates())

	# Divide a point by a number
	def divide(self, divisor: float):
		self.coordinates = [x/divisor for x in self.coordinates]

	# Returns a copy of the coordinates list from a point
	def get_coordinates(self, is_copy: bool = False) -> list:
		return self.coordinates.copy() if is_copy else self.coordinates

	# Sets a point's coordinates from a given list of floats
	def set_coordinates(self, new_coordinates: list):
		self.coordinates = new_coordinates


class Cluster:

	centroid: Point
	sum_of_points: Point
	number_of_points = 0
	is_smaller_than_epsilon = False

	def __init__(self, init_value: Point):
		self.centroid = Point(coordinates=init_value.get_coordinates(is_copy=True))  # Generates a centroid from a given point

		# Initializes sum_of_points to an empty Point
		dimension = len(self.centroid.get_coordinates())
		self.sum_of_points = Point(coordinates=([0.0] * dimension))

	def __repr__(self):
		return self.centroid.__repr__()

	def __str__(self):
		return self.centroid.__str__()

	# Update a cluster's centroid given the cluster's sum and number of points (calcs the mean of this cluster)
	# Returns a float representing the difference between the previous centroid and the new one.
	def update_centroid(self) -> float:
		old_centroid = Point(coordinates=self.centroid.get_coordinates())
		self.centroid.set_coordinates(new_coordinates=self.sum_of_points.get_coordinates(is_copy=True))
		self.centroid.divide(divisor=self.number_of_points)

		centroid_diff = self.centroid.distance_to_point(other_point=old_centroid)

		return centroid_diff

	# Removes an existing point from a cluster
	def remove_point(self, point: Point):
		self.sum_of_points -= point
		self.number_of_points -= 1

	# Adds a new point to a cluster
	def add_point(self, point: Point):
		self.sum_of_points += point
		self.number_of_points += 1

	# Calculates the distance between the cluster's centroid (which defines it) and a given point
	def calc_distance_from_point(self, point: Point) -> float:
		return point.distance_to_point(other_point=self.centroid)


class Kmeans:

	clusters: list
	datapoints: list
	max_iter: int
	k: int

	def __repr__(self):
		kmeans_description =\
			f"K: {self.k}, " \
			f"No. of Datapoints: {len(self.datapoints)}, " \
			f"Iter: {self.max_iter}, " \
			f"No. of Clusters: {len(self.clusters)}, " \
			f"Clusters: {self.clusters}"
		return kmeans_description

	def __str__(self):

		kmeans_description = f"""
			K: {self.k}, 
			No. of Datapoints: {len(self.datapoints)}, 
			Iter: {self.max_iter}, 
			No. of Clusters: {len(self.clusters)}, 
			Clusters: {self.clusters}, 
		"""

		return kmeans_description

	@staticmethod
	def __is_natural(n) -> bool:
		return str(n).isdigit() and float(n) == int(n) and int(n) > 0

	# Reader for the data.txt file
	@staticmethod
	def get_points_from_data(input_file_path: str) -> list:
		points = []
		with open(input_file_path, 'r') as data:
			for line in data:
				if line:  # Avoid blank lines
					point_coordinates = [float(x) for x in line.split(',')]
					points.append(Point(coordinates=point_coordinates))
		return points

	def __init__(self, k: int, data_file_url: str, max_iter: int):

		self.datapoints = self.get_points_from_data(input_file_path=data_file_url)

		assert self.__is_natural(k) and 1 < k < len(self.datapoints), "Invalid number of clusters!"
		assert self.__is_natural(max_iter) and 1 < max_iter < MAX_ITER, "Invalid maximum iteration!"

		self.max_iter = max_iter
		self.k = k
		self.clusters = list()

	# Initializes the k clusters from the first k datapoints
	def init_clusters(self):
		for i in range(self.k):
			init_point = Point(coordinates=self.datapoints[i].get_coordinates())
			self.clusters.append(Cluster(init_value=init_point))

	# The output for the assignment
	def print_clusters(self):
		for cluster in self.clusters:
			print(cluster)

	# Bulk calculates the distances from a given point to all k clusters.
	# Returns a list of tuples, each with the following format:
	# 	(distance_from_cluster_to_point, cluster)
	def calc_distances_from_point_to_all_clusters(self, point: Point) -> list:
		return [(cluster.calc_distance_from_point(point=point), cluster) for cluster in self.clusters]

	# Returns the nearest cluster from a given point
	def find_nearest_cluster_to_point(self, point: Point) -> Cluster:

		distance_cluster_tuples_list = self.calc_distances_from_point_to_all_clusters(point=point)

		# Returns the second element of the tuple (a cluster)
		# which its distance to the given point is the minimum ('key' is set to address only the distance parameter)
		return min(distance_cluster_tuples_list, key=lambda t: t[0])[1]

	def run_kmeans(self):

		# Init k clusters
		self.init_clusters()

		# Flags for termination conditions
		iteration_number = 0
		done_clusters = 0

		while (iteration_number < self.max_iter) and (done_clusters < self.k):
			done_clusters = 0

			# For each datapoint
			for point in self.datapoints:

				# Find its nearest cluster
				nearest_cluster = self.find_nearest_cluster_to_point(point=point)

				# Assign the datapoint to that cluster (unassign from another cluster if needed)
				point.assign_to_cluster(cluster=nearest_cluster)

			# Update clusters' centroid value per this iteration
			for cluster in self.clusters:

				# Calculate the size of change done on this iteration
				centroid_diff = cluster.update_centroid()

				# If the change is small enough, flag (count) it
				if centroid_diff < EPSILON:
					done_clusters += 1
			iteration_number += 1
		self.print_clusters()


if __name__ == '__main__':
	k = int(sys.argv[1])
	max_iter = int(sys.argv[2]) if len(sys.argv) == 4 else DEFAULT_ITER
	data_url = sys.argv[-1]

	kmeans_runner = Kmeans(k=k, data_file_url=data_url, max_iter=max_iter)
	kmeans_runner.run_kmeans()
